examples/step-86/step-86.cc

```diff
@@ -0,0 +1,1138 @@
+    // difference method will produce the correct result. This can be activated
+    // at run time via passing the `-snes_test_jacobian` option on the command
+    // line.
+    for (const auto &c : current_constraints.get_lines())
+      jacobian_matrix.set(c.index, c.index, 1.0);
+    jacobian_matrix.compress(VectorOperation::insert);
+  }
+
+
+  // @sect4{The HeatEquation::solve_with_jacobian() function}
+  //
+  // This is the function that actually solves the linear system with the
+  // Jacobian matrix we have previously built (in a call to the previous
+  // function during the current time step or another earlier one -- time
+  // steppers are quite sophisticated in determining internally whether it is
+  // necessary to update the Jacobian matrix, or whether one can reuse it for
+  // another time step without rebuilding it; this is similar to how one can
+  // re-use the Newton matrix for several Newton steps, see for example the
+  // discussion in step-77). We could in principle not provide this function to
+  // the time stepper, and instead select a specific solver on the command line
+  // by using the `-ksp_*` options of PETSc. However, by providing this
+  // function, we can use a specific solver and preconditioner for the linear
+  // system, and still have the possibility to change them on the command line.
+  //
+  // Providing a specific solver is more in line with the way we usually do
+  // things in other deal.II examples, while letting PETSc choose a generic
+  // solver, and changing it on the command line via `-ksp_type` is more in line
+  // with the way PETSc is usually used. Both options are available here, since
+  // we can still change both the solver and the preconditioner on the command
+  // line via `-user_ksp_type` and `-user_pc_type` options.
+  //
+  // In any case, recall that the Jacobian we built in the previous function is
+  // always of the form
+  // @f[
+  //   J_\alpha = \alpha M + A
+  // @f]
+  // where $M$ is a mass matrix and $A$ a Laplace matrix. $M$ is symmetric and
+  // positive definite; $A$ is symmetric and at least positive semidefinite;
+  // $\alpha> 0$. As a consequence, the Jacobian matrix is a symmetric and
+  // positive definite matrix, which we can efficiently solve with the Conjugate
+  // Gradient method, along with either SSOR or (if available) the algebraic
+  // multigrid implementation provided by PETSc (via the Hypre package) as
+  // preconditioner.
+  template <int dim>
+  void
+  HeatEquation<dim>::solve_with_jacobian(const PETScWrappers::MPI::Vector &src,
+                                         PETScWrappers::MPI::Vector       &dst)
+  {
+    TimerOutput::Scope t(computing_timer, "solve with Jacobian");
+
+#if defined(PETSC_HAVE_HYPRE)
+    PETScWrappers::PreconditionBoomerAMG preconditioner;
+    preconditioner.initialize(jacobian_matrix);
+#else
+    PETScWrappers::PreconditionSSOR preconditioner;
```

Comment:
PETSc has his native AMG ('-pc_type gamg'). It would make sense to use it here (but it is not wrapped), since SSOR is a bad choice.  Missing deal.II interfaces to PETSc solvers is why I prefer command line options solvers. Once the optimal is found, then it makes sense to add a `solve_with_jacobian callback`

Reply:
Yes, fair point. I augmented the comment in front of the function.